app/javascript/components/util/EligibilityTooltip.js

```diff
@@ -24,6 +24,12 @@ class EligibilityTooltip extends React.Component {
     };
   }
 
+  static getDerivedStateFromProps(nextProps) {
+    let activeOption = eligibility_options.find(eo => nextProps.report_eligibility[`${eo.conditional}`]) || _.last(eligibility_options);
+    activeOption.messages = activeOption.messageCount === 1 ? [nextProps.report_eligibility.messages[0]] : nextProps.report_eligibility.messages;
+    return { activeOption };
+  }
```

Comment:
I'm admittedly struggling a bit to understand the bug and why this component wasn't updating properly. Can you explain a little more?

Also the docs for [`getDerivedStateFromProps`](https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops) indicate that it is for rare use cases. Can you comment a little more on why using is necessary in this case, as opposed to allegedly [more common ways](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html)?

Reply:
The component wouldn't update, because `activeOption` is only set in the `constructor`. If just the `props` of this component update, we don't re-run the `constructor`, so the `activeOption` is not updated. You can actually see this if you test out recreating this bug. If you inspect some `EligibilityTooltip` components after applying an advanced filter on `master`, you will see the `props` look correct, but the `state` will be wrong. And since the `state` is what is used to render, we get the wrong icon. Let me know if that makes sense.

In our case we are effectively using the `state` for memoization. According to the [docs](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#what-about-memoization) this is one of the less bad uses for derived state. I can't exactly speak to the initial implementation, but I'm guessing the idea was to only calculate `activeOption` when it needed to be calculated. So I was trying to stick with that idea.

But now that I look at this more, would you ever re-render without wanting to recalculate `activeOption`? I'm thinking we probably could just move the calc of `activeOption` into render and remove it as state, since I wouldn't expect any of the other props on this component to be changing.

Reply:
703a08e updates this to just get rid of state altogether. It should still fix the bug, and as far as I can understand shouldn't be any less efficient in practice. Let me know what you think @tstrass.

Reply:
Ah I follow now, thanks for the thorough explanation. New implementation looks great to me!